src/contracts/core/StakeRootCompendium.sol

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.12;
+
+import "../interfaces/IAVSDirectory.sol";
+import "../interfaces/IDelegationManager.sol";
+import "../interfaces/IStakeRootCompendium.sol";
+import "../interfaces/IStrategy.sol";
+import "../libraries/Merkle.sol";
+
+import "@risc0-ethereum/contracts/src/IRiscZeroVerifier.sol";
+import "@openzeppelin-upgrades/contracts/access/OwnableUpgradeable.sol";
+
+
+contract StakeRootCompendium is IStakeRootCompendium, OwnableUpgradeable {
+
+    uint32 public constant MAX_OPERATOR_SET_SIZE = 2048;
+    uint32 public constant MAX_NUM_OPERATOR_SETS = 2048;
+    uint32 public constant MAX_NUM_STRATEGIES = 16;
+
+    IDelegationManager public delegation;
+    IAVSDirectory public avsDirectory;
+
+    address public verifier;
+    bytes32 public imageId;
+
+    mapping(uint32 => mapping(IStrategy => uint96)) public operatorSetIdToStrategyToMultiplier;
+
+    modifier isOperatorSetAVS(address avs) {
+        require(avsDirectory.isOperatorSetAVS(avs), "StakeRootCompendium: is not operator set AVS");
+        _;
+    }
+
+    constructor(
+        IDelegationManager _delegation,
+        IAVSDirectory _avsDirectory
+    ) {
+        _disableInitializers();
+        delegation = _delegation;
+        avsDirectory = _avsDirectory;
+    }
+
+     function getStakeRoot(address[] calldata avss, bytes32[] calldata operatorSetRoots) external view returns (bytes32) {
+        require(avss.length == operatorSetRoots.length, "AVSSyncTree.getStakeRoot: AVS and operator set roots length mismatch");
+        require(operatorSetRoots.length <= MAX_NUM_OPERATOR_SETS, "AVSSyncTree.getStakeRoot: operatorSet size limit exceeded");
+    
+        bytes32[] memory operatorSetLeaves = new bytes32[](avss.length);
+        for (uint256 i = 0; i < avss.length; i++) {
+            operatorSetLeaves[i] = keccak256(abi.encodePacked(avss[i], operatorSetRoots[i]));
+        }
+        return Merkle.merkleizeKeccak256(operatorSetLeaves);
+    }
+
+    function getOperatorSetRoot(address avs, uint32 operatorSetId, address[] calldata operators, IStrategy[] calldata strategies) external view isOperatorSetAVS(avs) returns (bytes32) {
+        require(operators.length <= MAX_OPERATOR_SET_SIZE, "AVSSyncTree._verifyOperatorStatus: operator set too large");
+
+        bytes32[] memory operatorLeaves = new bytes32[](operators.length);
+        for (uint256 i = 0; i < operators.length; i++) {
+            require(delegation.isOperator(operators[i]), "AVSSyncTree.getOperatorSetRoot: operator not registered");
+            require(avsDirectory.avsOperatorStatus(avs,operators[i]) == IAVSDirectory.OperatorAVSRegistrationStatus.REGISTERED, "AVSSyncTree.getOperatorSetRoot: operator not registered to AVS");
+            require(avsDirectory.isMember(avs,operators[i],operatorSetId), "AVSSyncTree.getOperatorSetRoot: operator not in operator set");
```

Comment:
this is the only check that should be needed, the other two are redundant (already checked to get to this state)

Reply:
added a note stating that assumption